app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,314 @@
+            // fetch needed data
+            val fetchBBox = tilesToFetch.minTileRect()!!.asBoundingBox(tileZoom)
+            val (elements, geometries) = fetchMapData(fetchBBox)
+
+            // get nodes from spatial cache
+            // this may not contain all nodes, but tiles that were cached initially might
+            // get dropped when the caches are updated
+            // duplicate fetch might be unnecessary often, but it's fast compared to fetching from db
+            nodes = HashSet<Node>().apply { addAll(spatialCache.get(bbox)) }
+
+            // put data to caches
+            update(addedOrUpdatedElements = elements, addedOrUpdatedGeometries = geometries, bbox = fetchBBox)
+
+            // return data if we need exactly what was just fetched
+            if (fetchBBox == bbox) {
+                result.putAll(elements, geometries + elements.filterIsInstance<Node>().map { it.toElementGeometryEntry() })
+                return result
+            }
+
+            // get nodes again, should contains the newly added nodes this time
+            // (but maybe not the old ones)
+            nodes.addAll(spatialCache.get(bbox))
+        } else {
+            nodes = spatialCache.get(bbox)
+        }
+
+        nodes.forEach { result.put(it, ElementPointGeometry(it.position)) } // create a new geometry, as it's not cached (this is the slowest part in loading from cache)
+        synchronized(this) {
+            val wayIds = HashSet<Long>(nodes.size / 5)
+            val relationIds = HashSet<Long>(nodes.size / 10)
+            nodes.forEach { node ->
+                wayIdsByNodeIdCache[node.id]?.let { wayIds.addAll(it) }
+                relationIdsByElementKeyCache[ElementKey(ElementType.NODE, node.id)]?.let { relationIds.addAll(it) }
+            }
+            wayIds.forEach { wayId ->
+                relationIdsByElementKeyCache[ElementKey(ElementType.WAY, wayId)]?.let { relationIds.addAll(it) }
+            }
+
+            wayIds.forEach {
+                val key = ElementKey(ElementType.WAY, it)
+                result.put(wayRelationCache[key]!!, wayRelationGeometryCache[key])
+            }
+            relationIds.forEach {
+                val key = ElementKey(ElementType.RELATION, it)
+                result.put(wayRelationCache[key]!!, wayRelationGeometryCache[key])
+            }
+        }
+
+        // todo: call trimNonSpatialCaches if tilesToFetch.isNotEmpty()?
+        //  we just added a lot of data to cache, so maybe we should drop something?
+        //  or maybe first check whether spatialCache is full?
+
+        return result
+    }
+
+    fun clear() {
+        synchronized(this) {
+            wayIdsByNodeIdCache.clear()
+            relationIdsByElementKeyCache.clear()
+            wayRelationCache.clear()
+            wayRelationGeometryCache.clear()
+            spatialCache.clear()
+        }
+    }
+
+    fun trim(size: Int) {
+        spatialCache.trim(size)
+        trimNonSpatialCaches()
```

Comment:
SpatialCache is synchronized anyway, so is it really necessary to synchronize access also in `MapDataCache`?

Accesses to `HashMap`s should be synchronized, e.g. in this case `trimNonSpatialCaches` is synchronized, but I may have overlooked something.

Reply:
(I think) my original intention was to keep the blocking by `synchronized` as short as possible. But I think you're right, better be safe and synchronize everything to avoid potential inconsistencies.